src/lib/yang/xpath.lua

```diff
@@ -0,0 +1,280 @@
+local lib = require("core.lib")
+local schemalib = require("lib.yang.schema")
+local datalib = require("lib.yang.data")
+local valuelib = require("lib.yang.value")
+local util = require("lib.yang.util")
+local normalize_id = datalib.normalize_id
+
+local function extract_parts(fragment)
+   local rtn = {query={}}
+   rtn.name = string.match(fragment, "([^%[]+)")
+   for k,v in string.gmatch(fragment, "%[([^=]+)=([^%]]+)%]") do
+      rtn.query[k] = v
+   end
+   return rtn
+end
+
+local handlers = {}
+function handlers.scalar(fragment, tree)
+   return {name=fragment.name}, tree
+end
+function handlers.struct(fragment, tree)
+   return {name=fragment.name}, tree.members
+end
+function handlers.table(fragment, tree)
+   return {name=fragment.name, keys=fragment.query}, tree
+end
+function handlers.array(fragment, tree)
+   local position = fragment.query["position()"]
+   return {name=fragment.name, key=tonumber(position)}
+end
+function handle(node_type, fragment, tree)
+   return assert(handlers[node_type], node_type)(fragment, tree)
+end
+
+-- Gets the next item in the path returning the element and the remaining
+-- path fragment. For example "router.routes.route" will return "router"
+-- and "routes.route". If the end is reached it'll return nil.
+local function next_element(path)
+   return string.match(path, "([^/]+)/?(.*)")
+end
+
+local function split_path(path)
+   local tail = path
+   return function ()
+      local head
+      head, tail = next_element(tail)
+      if head == nil then return head else return extract_parts(head) end
+   end
+end
+
+-- Finds the grammar node for a fragment in a given grammar.
+local function extract_grammar_node(grammar, fragment)
+   assert(grammar.type ~= "scalar")
+   local errmsg = "Invalid path: "..fragment.name
+   if grammar.type == "table" then
+      if grammar.keys[fragment.name] == nil then
+         return assert(grammar.values[fragment.name], errmsg)
+      else
+         return grammar.keys[fragment.name]
+      end
+   else
+      return assert(grammar[fragment.name], errmsg)
+   end
+end
+
+-- Converts an XPath path to a lua array consisting of path componants.
+-- A path component can then be resolved on a yang data tree:
+function convert_path(grammar, path)
+   local ret = {}
+   local node = grammar
+   for element in split_path(path) do
+      node = extract_grammar_node(node, element)
+      local luapath, next_node = handle(node.type, element, node)
+      luapath.grammar = node
+      table.insert(ret, luapath)
+      node = next_node
+   end
+   return ret
+end
+
+-- Returns a resolver for a paticular schema and *lua* path.
+function resolver(schema, path)
+   local grammar = datalib.data_grammar_from_schema(schema)
+   local handlers = {}
+   local function handle(scm, prod, data, path)
+      if #path == 0 then return data end
+      return assert(handlers[scm.kind], scm.kind)(scm, prod, data, path)
+   end
+   function handlers.list(scm, prod, data, path)
```

Comment:
I don't think we should be interpreting the schema here -- instead we should be building a resolver based on the grammar.  The grammar describes the concrete data format whereas the schema is at a higher level (e.g. presence), and the path should be in terms of data/grammar, not schema.  This is reflected in that the YANG XPath expressions in RFC 6020 apply (AFAIU) to the XML data, which follows the XML mapping rules.  (For us the grammar is like what happens to the schema after you apply XML mapping rules).  So we should have cases for tables, scalars, and so on, not containers, leaves, etc.  WDYT?